Bluetooth: Controller: Separate Tx/Rx frame spacing

cvinayak opened the pull request:
Separate Tx/Rx frame spacing.

Blocked by defines in #79584

cvinayak committed:
Bluetooth: Controller: 1 ms connection

1 ms connection interval support.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
tests: bsim: Bluetooth: Test 1ms connection interval support

Test 1ms connection interval support.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Separate Tx/Rx frame spacing

Separate Tx/Rx frame spacing.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Add radio timer ISR usage support

Add radio timer ISR usage support.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Minor updates to code comments

Minor updates and correction to code comments.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Introduce deferred ACL Tx packet transmission

Introduce deferred ACL Tx packet transmission setup by radio,
so that an enqueued ACL Data packet can be transmitted with
the shortest latency.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
tests: bsim: Bluetooth: Deferred ACL Tx transmission setup

Added test to verify deferred ACL Tx transmission setup,
such that GATT discovery completes with least latency and
GATT notifications can be received earliest after connection
establishment.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented:
@Thalley any idea what happened in the bsim nrf5340 testing of unicast client in this PR?

Thalley commented:
> @Thalley any idea what happened in the bsim nrf5340 testing of unicast client in this PR?

Looks like one device gets a segmentation fault which triggers a CIS disconnect. The client, attempting to send, does not get the callback that the stream is stopped until after a few TX attempts it seems

cvinayak commented:
> > @Thalley any idea what happened in the bsim nrf5340 testing of unicast client in this PR?
> 
> Looks like one device gets a segmentation fault which triggers a CIS disconnect. The client, attempting to send, does not get the callback that the stream is stopped until after a few TX attempts it seems

ok... it does not say which, app core or net core crashes... will try running at desk to determine the cause.

cvinayak commented:
> > > @Thalley any idea what happened in the bsim nrf5340 testing of unicast client in this PR?
> > 
> > 
> > Looks like one device gets a segmentation fault which triggers a CIS disconnect. The client, attempting to send, does not get the callback that the stream is stopped until after a few TX attempts it seems
> 
> ok... it does not say which, app core or net core crashes... will try running at desk to determine the cause.

The segmentation fault is due to having `net_buf_alloc()` call (which is no more has the blocking behavior) in the `audio_timer_timeout()` system work queue work in the `samples/bluetooth/bap_unicast_server/src/main.c` .

@Thalley The audio samples will need to be ported to gracefully handle `net_buf_alloc()` returning `NULL` if being called inside system work queue...

Thalley commented:
> > > > @Thalley any idea what happened in the bsim nrf5340 testing of unicast client in this PR?
> > > 
> > > 
> > > Looks like one device gets a segmentation fault which triggers a CIS disconnect. The client, attempting to send, does not get the callback that the stream is stopped until after a few TX attempts it seems
> > 
> > 
> > ok... it does not say which, app core or net core crashes... will try running at desk to determine the cause.
> 
> The segmentation fault is due to having `net_buf_alloc()` call (which is no more has the blocking behavior) in the `audio_timer_timeout()` system work queue work in the `samples/bluetooth/bap_unicast_server/src/main.c` .
> 
> @Thalley The audio samples will need to be ported to gracefully handle `net_buf_alloc()` returning `NULL` if being called inside system work queue...

Huh, weird, thought I had fixed all the `K_FOREVER` in the samples's workqueue callbacks. Will fix today

Thalley commented:
@cvinayak https://github.com/zephyrproject-rtos/zephyr/pull/79187 should fix the remaining samples.

Thalley commented on subsys/bluetooth/controller/ll_sw/pdu.h:
```diff
@@ -185,6 +185,10 @@
 #define EVENT_CLOCK_JITTER_US   2
 /* Event interframe timings */
 #define EVENT_IFS_US            150
+/* Event interframe timings, low latency mode */
+#define EVENT_IFS_LOW_LAT_US    (CONFIG_BT_CTLR_IFS_LOW_LATENCY)
```
What is the rationale behind adding this define that is defined to be the same as a Kconfig, rather than using the Kconfig value directly? 

IMO the latter may provide more information to the reader in terms of knowing that it's a configurable value and not a static value

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -698,7 +698,13 @@ uint8_t ll_apto_get(uint16_t handle, uint16_t *apto)
 	}
 
-	*apto = conn->apto_reload * conn->lll.interval * 125U / 1000;
+	if (conn->lll.interval > 5U) {
+		*apto = conn->apto_reload * conn->lll.interval *
+			CONN_INT_UNIT_US / 10000U;
+	} else {
+		*apto = conn->apto_reload * (conn->lll.interval + 1U) *
+			CONN_LOW_LAT_INT_UNIT_US / 10000U;
+	}
```
I think there are some magic numbers (`5U` and `10000U` that are either defined or should be defined. The division by `10000U` for a microsecond value would appears to be a translation to a different unit, and could be a macro or something.

cvinayak replied:
Host uses hardcoded value `6` and i could not easily find a define for minimum BLE Connection Interval!  The timeout are in 10 ms unit, same here, do we have those defines?

Thalley replied:
Yeah, for some reason they were not defined before https://github.com/zephyrproject-rtos/zephyr/pull/79584 

I don't think we have any conversion macros between ms and units of 10ms, so we usually just do it like here, but with a comment stating that we are converting to a centiseconds (10ms). 

It could help to replace `10000` with `USEC_PER_MSEC * 10` as well

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -712,9 +718,17 @@ uint8_t ll_apto_set(uint16_t handle, uint16_t apto)
-	conn->apto_reload = RADIO_CONN_EVENTS(apto * 10U * 1000U,
-					      conn->lll.interval *
-					      CONN_INT_UNIT_US);
+	if (conn->lll.interval > 5U) {
+		conn->apto_reload =
+			RADIO_CONN_EVENTS(apto * 10U * 1000U,
+					  conn->lll.interval *
+					  CONN_INT_UNIT_US);
+	} else {
+		conn->apto_reload =
+			RADIO_CONN_EVENTS(apto * 10U * 1000U,
+					  (conn->lll.interval + 1U) *
+					  CONN_LOW_LAT_INT_UNIT_US);
+	}
```
Ditto with magic numbers

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -1075,7 +1098,15 @@ void ull_conn_done(struct node_rx_event_done *done)
-			const uint32_t conn_interval_us = conn->lll.interval * CONN_INT_UNIT_US;
+			uint32_t conn_interval_us;
+
+			if (conn->lll.interval > 5U) {
```
The value `5U` here for interval is also magic.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -2165,16 +2201,96 @@ void ull_conn_update_parameters(struct ll_conn *conn, uint8_t is_cu_proc, uint8_
+		conn_interval_unit_old = CONN_LOW_LAT_INT_UNIT_US;
+	}
+
+	if (interval > 5U) {
```
Ditto

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -2165,16 +2201,96 @@ void ull_conn_update_parameters(struct ll_conn *conn, uint8_t is_cu_proc, uint8_
+					    PHY_1M);
+
+#if defined(CONFIG_BT_CTLR_PHY)
+		max_tx_time = MAX(max_tx_time,
+				  PDU_DC_MAX_US(PDU_DC_PAYLOAD_SIZE_MIN,
+					 lll->phy_tx));
+		max_rx_time = MAX(max_rx_time,
+				  PDU_DC_MAX_US(PDU_DC_PAYLOAD_SIZE_MIN,
+					 lll->phy_rx));
```
```suggestion
		max_tx_time = MAX(max_tx_time,
				  PDU_DC_MAX_US(PDU_DC_PAYLOAD_SIZE_MIN,
						lll->phy_tx));
		max_rx_time = MAX(max_rx_time,
				  PDU_DC_MAX_US(PDU_DC_PAYLOAD_SIZE_MIN,
						lll->phy_rx));
```

cvinayak replied:
done

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -2165,16 +2201,96 @@ void ull_conn_update_parameters(struct ll_conn *conn, uint8_t is_cu_proc, uint8_
+		slot_us += ready_delay_us;
+
+		if (IS_ENABLED(CONFIG_BT_CTLR_EVENT_OVERHEAD_RESERVE_MAX) ||
+		    !lll->role) {
```
```suggestion
		    lll->role == BT_HCI_ROLE_CENTRAL) {
```

cvinayak replied:
done

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -2165,16 +2201,96 @@ void ull_conn_update_parameters(struct ll_conn *conn, uint8_t is_cu_proc, uint8_
+
+	conn_interval_old_us = conn_interval_old *
+			       conn_interval_unit_old;
+	latency_upd = conn_interval_old_us / (conn_interval_new *
+					  conn_interval_unit_new);
```
```suggestion
	latency_upd = conn_interval_old_us / (conn_interval_new *
					      conn_interval_unit_new);
```

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -2165,16 +2201,96 @@ void ull_conn_update_parameters(struct ll_conn *conn, uint8_t is_cu_proc, uint8_
+			       conn_interval_unit_old;
+	latency_upd = conn_interval_old_us / (conn_interval_new *
+					  conn_interval_unit_new);
+	conn_interval_new_us = latency_upd * conn_interval_new *
+			       conn_interval_unit_new;
```
The `conn_interval_new * conn_interval_unit_new` is done twice; consider storing the result in a temp variable

Thalley replied:
It's actually done thrice as we have `conn_interval_us = conn_interval_new * conn_interval_unit_new;` further below

cvinayak replied:
good catch.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -2278,7 +2396,13 @@ void ull_conn_update_peer_sca(struct ll_conn *conn)
 
 	/* calculate the window widening and interval */
-	conn_interval_us = lll->interval * CONN_INT_UNIT_US;
+	if (lll->interval > 5U) {
```
Ditto

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_internal.h:
```diff
@@ -8,11 +8,15 @@
  */
 #if !defined(CONFIG_BT_CTLR_USER_CPR_INTERVAL_MIN)
+#if defined(CONFIG_BT_CTLR_CONN_INTERVAL_LOW_LATENCY)
+#define CONN_INTERVAL_MIN(x) (0U)
```
Is `0` right here? Shouldn't it be `1`?

Thalley replied:
Hmm, `0` means `500us` sounds weird to me :D If `ms` isn't possible to use, then perhaps these macros should be in `us` instead? It sounds weird that an interval of 0 is not in fact 0.

Thalley commented on tests/bsim/bluetooth/ll/conn/src/test_connect2.c:
```diff
@@ -209,7 +209,12 @@ static void test_con2_main(void)
 	 */
 	while (1) {
-		k_sleep(K_SECONDS(1));
+		if (IS_ENABLED(CONFIG_TEST_CONN_INTERVAL_1MS) ||
+		    IS_ENABLED(CONFIG_BT_CTLR_TX_DEFER)) {
+			k_sleep(K_MSEC(1));
+		} else {
+			k_sleep(K_SECONDS(1));
+		}
```
Hmm, this looks kind of odd. I guess it makes sense to send a notification every 1ms when testing with 1ms conn interval, but why are we sending every 1 second if not? What if the interval is 10ms, shouldn't we send every 10ms then? What if the connection interval is 4 seconds?

Shouldn't we just ignore the Kconfig values and just always sleep for the current connection interval?

cvinayak replied:
The original (probably the very first) bsim test simulated a peripheral HR service where the notification were 1 second apart. Do not want to change the existing test behavior in this PR.

Only the verification that the number of notifications received in the 3 seconds of simulation has been updated.

Thalley replied:
> Do not want to change the existing test behavior in this PR.

We can create a PR modifying the behavior to use the connection interval first, and then you wouldn't need this change at all ;)

Thalley commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -1049,6 +1064,19 @@ config BT_CTLR_ALLOW_SAME_PEER_CONN
 	  WARNING: This option enables behavior that violates the Bluetooth
 	  specification.
 
+config BT_CTLR_CONN_INTERVAL_LOW_LATENCY
+	bool "Allow low latency connection intervals"
+	help
+	  Allow low latency connection intervals.
```
How is this expected to be used with the Zephyr host?

IIRC we test for valid HCI values (e.g. values under 7.5 cannot be requested), and if the connection completed event contain values lower than what's allowed with HCI, then that's considered invalid as well.

Thalley replied:
Ah, forgot about that one :)

cvinayak committed:
Bluetooth: Controller: Defines for radio timer capture/compare indices

Add defines for radio timer capture and compare indices used
in the nRF Radio HAL implementation.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -1153,6 +1153,55 @@ uint32_t radio_bc_has_match(void)
+
+		/* Setup compare event with min. 1 us offset */
+		nrf_timer_event_clear(EVENT_TIMER, NRF_TIMER_EVENT_COMPARE2);
+		nrf_timer_cc_set(EVENT_TIMER, 2, start_us + 1U);
```
Added defines for hard coded register indices.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c:
```diff
@@ -1153,6 +1153,55 @@ uint32_t radio_bc_has_match(void)
+				       HAL_EVENT_TIMER_SAMPLE_TASK);
+
+		now_us = EVENT_TIMER->CC[HAL_EVENT_TIMER_SAMPLE_CC_OFFSET];
+	} while ((now_us > start_us) && (EVENT_TIMER->EVENTS_COMPARE[2] == 0U));
```
Added defines.

Thalley commented on subsys/bluetooth/controller/ll_sw/pdu.h:
```diff
@@ -185,6 +185,8 @@
 #define EVENT_CLOCK_JITTER_US   2
 /* Event interframe timings */
 #define EVENT_IFS_US            150
+/* Event interframe timings, default */
+#define EVENT_IFS_DEFAULT_US    EVENT_IFS_US
 /* Standard allows 2 us timing uncertainty inside the event */
 #define EVENT_IFS_MAX_US        (EVENT_IFS_US + EVENT_CLOCK_JITTER_US)
```
What's the purpose of the duplicated `#define`? 

If we intend to keep it, should that be used for `EVENT_IFS_MAX_US` since that refers to a "standard"? 

Will these 2 ever be different?

cvinayak replied:
Looks like there is inconsistent use of `EVENT_IFS_US`, `EVENT_IFS_MAX_US`, `+ 4, `- 4` and `EVENT_CLOCK_JITTER_US`  .... this needs clean, but out side the scope of this PR.

Thalley reviewed (approved):
As usual, my reviews for the controller does not verify the correctness of the implementation as I don't know all the details, but CI is passing and the changes look OK to me
